### DISCUSSION TITLE
How can I get the exit code from pods().exec()?

### DISCUSSION THREAD

**Discussion by ghost** [fabric8io/kubernetes-client#4040] [Q&A]
I've use the example to run a command in a pod, and the command is executing successfully. However, I would like to capture the exit code, and I don't see any API method that allows me to do that.

```shell
$ kubectl exec -it mysql-0 -- true; echo $?
0
$ kubectl exec -it mysql-0 -- false; echo $?
command terminated with exit code 1
1
```

My Java code:
```java
        SimpleExecListener listener = new SimpleExecListener();
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        ByteArrayOutputStream err = new ByteArrayOutputStream();
        getKubernetesClient().pods().inNamespace(getNamespace()).withName(resource)
                .inContainer("solr")
                .readingInput(InputStream.nullInputStream())
                .writingOutput(out)
                .writingError(err)
                .usingListener(listener)
                .exec("/bin/sh", "-c", "echo foo");
        listener.awaitUninterruptable();
```

My SimpleListener captures the code from the `onClose()` method, but that is always 1000, and looking at the code in `ExecWebSocketListener`, it seems to be something from the WebSocket client, not the exit code of the executed process. I would have expected `ExecWatch` to have a method to retrieve the code, but it doesn't.

**Comment by ghost** [comment]
This is pretty much what I found in the example: https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/ExecExample.java

**Reply by ghost** [reply]
When I exec false, the flow remains the same; it seems onFailure is called only when the exec fails, not when the command starts and exits with a non-zero exit code (which is fine, btw, just I'd like to get the exit code...)

**Comment by shawkins** [comment]
#3813 added ExecListener.onExit.  And a recent change for 6.0 added ExecWatch.exitCode.

**Comment by ghost** [comment]
Ah, thanks, I really need to upgrade to 6.

**Reply by shawkins** [reply]
There isn't a release date set yet unfortunately.

**Comment by manusa** [comment]
For completeness:

Version 6.x is out and there is an [`ExecWatch.exitCode()`](https://github.com/fabric8io/kubernetes-client/blob/de7e27a562e31b0b34b9c1592bc3aeedfb262bdb/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/dsl/ExecWatch.java#L82) method that returns a `CompletableFuture` that returns an Integer with the exit code when completed.
